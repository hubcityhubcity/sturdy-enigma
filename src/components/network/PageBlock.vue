<template>
  <TmPage
    data-title="Block"
    :managed="true"
    :loading="this.$apollo.queries.block.loading"
    :error="error"
    hide-header
  >
    <template slot="managed-body">
      <div class="block">
        <h2 class="page-profile__title">Block {{ height }}</h2>
      </div>

      <ul class="row">
        <li>
          <h4>Chain ID</h4>
          <span class="page-data">{{ block.chainId }}</span>
        </li>
        <li>
          <h4>Time</h4>
          <span class="page-data">{{ block.time | date }}</span>
        </li>
      </ul>

      <div class="row">
        <div class="column">
          <h3 class="page-profile__section-title">
            Transactions ({{ block.transactions.length }})
          </h3>

          <TmDataMsg v-if="block.transactions.length === 0" icon="info_outline">
            <div slot="title">
              No Transactions
            </div>
            <div slot="subtitle">
              This block doesn't contain any transactions.
            </div>
          </TmDataMsg>

          <TransactionList
            :transactions="block.transactions"
            :address="address"
            :validators="validatorsAddressMap"
          />
          <br />
        </div>
      </div>
    </template>
  </TmPage>
</template>

<script>
import { mapGetters } from "vuex"
import { date } from "src/filters"
import gql from "graphql-tag"

import TmPage from "common/TmPage"
import TransactionList from "transactions/TransactionList"
import TmDataMsg from "common/TmDataMsg"
export default {
  name: `page-block`,
  components: {
    TmDataMsg,
    TmPage,
    TransactionList
  },
  filters: {
    date
  },
  data: () => ({
    block: {
      transactions: []
    },
    validators: [],
    error: undefined
  }),
  computed: {
    ...mapGetters([`address`, `network`]),
    height() {
      return this.$route.params.height
    },
    validatorsAddressMap() {
      const names = {}
      this.validators.forEach(item => {
        names[item.operator_address] = item
      })
      return names
    }
  },
  apollo: {
    validators: {
      query: gql`
        query validators(
          $networkId: String!
          $delegatorAddress: String
          $all: Boolean
          $query: String
        ) {
          validators(
            networkId: $networkId
            delegatorAddress: $delegatorAddress
            all: $all
            query: $query
          ) {
            name
            operatorAddress
          }
        }
      `,
      variables() {
        return {
          networkId: this.network
        }
      },
      update: function(result) {
        return result.validators
      }
    },
    block: {
      query: gql`
        query block($networkId: String!, $height: Int!) {
          block(networkId: $networkId, height: $height) {
            networkId
            height
            hash
            chainId
            time
            transactions {
              type
              hash
              height
              group
              timestamp
              gasUsed
              gasWanted
              success
              log
              memo
              fee {
                amount
                denom
              }
              signature
              value
              amount
            }
            proposer_address
          }
        }
      `,
      variables() {
        return {
          networkId: this.network,
          height: Number(this.height)
        }
      },
      update: function(result) {
        if (!result.block) {
          return {
            transactions: []
          }
        }

        const block = {
          ...result.block,
          transactions: result.block.transactions.map(transaction => ({
            ...transaction,
            timestamp: new Date(transaction.timestamp),
            value: JSON.parse(transaction.value)
          }))
        }
        return block
      },
      result({ error }) {
        // TODO move logic of 404 into API
        this.error = error
      }
    }
  }
}
</script>
<style scoped>
.page-profile__title {
  color: var(--bright);
  font-size: var(--h1);
  line-height: 2.25rem;
  font-weight: 500;
  padding: 1rem;
}

.page-profile__section-title {
  color: var(--txt);
}

.tm-data-msg {
  margin: 0;
}

.page-profile__section {
  padding-top: 2rem;
}
</style>
