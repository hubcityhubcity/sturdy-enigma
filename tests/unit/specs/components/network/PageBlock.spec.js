import { shallowMount, createLocalVue } from "@vue/test-utils"
import PageBlock from "network/PageBlock"
import { bankTxs } from "../../store/json/txs"

const localVue = createLocalVue()
localVue.directive(`tooltip`, () => { })

describe(`PageBlock`, () => {
  let wrapper

  const getters = {
    connected: true,
    lastHeader: {
      height: `1000`
    },
    validators: {},
    connection: {
      network: "testnet"
    }
  }

  const state = {
    delegation: {
      unbondingDelegations: {},
      loaded: true
    },
    session: { address: `` }
  }

  beforeEach(async () => {
    wrapper = shallowMount(PageBlock, {
      localVue,
      mocks: {
        $store: {
          getters,
          state,
          dispatch: jest.fn(type => {
            if (type === "queryBlockInfo")
              return {
                header: {
                  chain_id: `chain-1`,
                  num_txs: 10
                },
                block_id: {
                  hash: `ABCD1234`
                }
              }
            if (type === "getBlockTxs") return bankTxs
          })
        },
        $route: {
          params: { height: `100` }
        },
        $router: {
          push: jest.fn()
        }
      },
      stubs: [`router-link`]
    })
    await wrapper.vm.getBlock()
  })

  it(`shows a page with information about a certain block`, () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it(`shows the block page with no txs`, () => {
    wrapper = shallowMount(PageBlock, {
      localVue,
      mocks: {
        $store: {
          getters: {
            connected: true,
            lastHeader: {
              height: `1000`
            },
            validators: {}
          },
          state,
          dispatch: jest.fn(type => {
            if (type === "queryBlockInfos")
              return {
                header: {
                  chain_id: `chain-1`
                },
                block_id: {
                  hash: `ABCD1234`
                }
              }
            if (type === "getBlockTxs") return []
          })
        },
        $route: {
          params: { height: `100` }
        },
        $router: {
          push: jest.fn()
        }
      },
      stubs: [`router-link`]
    })

    expect(wrapper.element).toMatchSnapshot()
  })

  it(`loads the block information when the route changes`, () => {
    const getBlock = jest.fn()
    PageBlock.watch[`$route.params.height`].call({ getBlock })

    expect(getBlock).toHaveBeenCalled()
  })

  it(`redirects to the 404 page if the block doesn't exist`, async () => {
    const routerPush = jest.fn()

    await PageBlock.methods.getBlock({
      $store: {
        dispatch: () => null // not returning a block when querying
      },
      $route: {
        params: {
          height: `42`
        }
      },
      $router: {
        push: routerPush
      },
      lastHeader: {
        height: `1`
      }
    })

    expect(routerPush).toHaveBeenCalledWith(`/404`)

    routerPush.mockClear()
    await PageBlock.methods.getBlock({
      $store: {
        dispatch: () => ({ x: 1 }) // return pseudo block
      },
      $route: {
        params: {
          height: 42
        }
      },
      $router: {
        push: routerPush
      },
      lastHeader: {
        height: `1000`
      }
    })

    expect(routerPush).not.toHaveBeenCalledWith(`/404`)
  })
})
