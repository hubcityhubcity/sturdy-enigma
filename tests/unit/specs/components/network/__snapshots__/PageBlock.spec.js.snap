// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageBlock shows a page with information about a certain block 1`] = `
<tmpage-stub
  data-title="Block"
  hideheader="true"
  managed="true"
  subtitle=""
  title=""
>
  <template>
    <div
      class="block"
    >
      <h2
        class="page-profile__title"
      >
        Block #100
      </h2>
    </div>
     
    <ul
      class="row"
    >
      <li>
        <h4>
          Chain ID
        </h4>
         
        <span
          class="page-data"
        >
          gaia-13006
        </span>
      </li>
       
      <li>
        <h4>
          Time
        </h4>
         
        <span
          class="page-data"
        >
          September 24th 2019, 13:03
        </span>
      </li>
    </ul>
     
    <div
      class="row"
    >
      <div
        class="column"
      >
        <h3
          class="page-profile__section-title"
        >
          
          Transactions (1)
          
          <!---->
        </h3>
         
        <!---->
         
        <transactionlist-stub
          address="cosmos1"
          transactions="[object Object]"
          validators="[object Object]"
        />
         
        <br />
      </div>
    </div>
  </template>
</tmpage-stub>
`;

exports[`PageBlock shows the block page with no txs 1`] = `
<tmpage-stub
  data-title="Block"
  hideheader="true"
  managed="true"
  subtitle=""
  title=""
>
  <template>
    <div
      class="block"
    >
      <h2
        class="page-profile__title"
      >
        Block #100
      </h2>
    </div>
     
    <ul
      class="row"
    >
      <li>
        <h4>
          Chain ID
        </h4>
         
        <span
          class="page-data"
        >
          gaia-13006
        </span>
      </li>
       
      <li>
        <h4>
          Time
        </h4>
         
        <span
          class="page-data"
        >
          September 24th 2019, 13:03
        </span>
      </li>
    </ul>
     
    <div
      class="row"
    >
      <div
        class="column"
      >
        <h3
          class="page-profile__section-title"
        >
          
          Transactions (0)
          
          <!---->
        </h3>
         
        <tmdatamsg-stub
          icon="info_outline"
          iconcolor="var(--warning)"
        >
          <div>
            
            No Transactions
          
          </div>
           
          <div>
            
            This block doesn't contain any transactions.
          
          </div>
        </tmdatamsg-stub>
         
        <transactionlist-stub
          address="cosmos1"
          transactions=""
          validators="[object Object]"
        />
         
        <br />
      </div>
    </div>
  </template>
</tmpage-stub>
`;
